diff --git a/kernel-cssu-2.6.28/debian/control b/kernel-power-2.6.28/debian/control
index ac43331..ec1431c 100644
--- a/kernel-cssu-2.6.28/debian/control
+++ b/kernel-cssu-2.6.28/debian/control
@@ -3,7 +3,7 @@ Section: utils
 Priority: optional
 Maintainer: Pali Rohár <pali.rohar@gmail.com>
 XSBC-Bugtracker: https://garage.maemo.org/tracker/?group_id=1528
-Build-Depends: debhelper (>= 4.0.0), quilt, sdk-fiasco-gen, libncurses5-dev, sharutils
+Build-Depends: debhelper (>= 4.0.0), quilt, libncurses5-dev, sharutils
 Build-Depends-Indep: bzip2
 Standards-Version: 3.8.0
 
diff --git a/kernel-cssu-2.6.28/debian/patches/series b/kernel-power-2.6.28/debian/patches/series
index f4f3d98..74a1598 100644
--- a/kernel-cssu-2.6.28/debian/patches/series
+++ b/kernel-cssu-2.6.28/debian/patches/series
@@ -6,6 +6,7 @@ nokia-20101501+0m5.diff
 nokia-20103103+0m5.diff
 #nokia-20103103+0m5_usb.diff
 2.6.28.10.diff
+madde.diff
 rx51_defconfig.diff
 unionfs-2.5.3.diff
 dm-loop.diff
diff --git a/kernel-cssu-2.6.28/debian/rules b/kernel-power-2.6.28/debian/rules
index 6510f9e..51515ac 100755
--- a/kernel-cssu-2.6.28/debian/rules
+++ b/kernel-cssu-2.6.28/debian/rules
@@ -1,6 +1,8 @@
 #!/usr/bin/make -f
 #
 
+export PATH := $(HOME)/.madde/0.7.48/toolchains/arm-2007q3-51sb6-gdb71-arm-none-linux-gnueabi_linux_x86_64/bin:$(PATH)
+
 RELEASE := $(shell dpkg-parsechangelog | sed -ne 's/^Version: 1:\(.*\)/\1/p' | sed 's/-10/.10-/')
 REVISION := $(shell echo "$(RELEASE)" | sed 's/.*\./\./')
 EXTRAVERSION := EXTRAVERSION=$(REVISION)
@@ -78,7 +80,10 @@ install-kernel:
 	dh_testdir
 	dh_testroot
 	mkdir -p $(CURDIR)/debian/$(KERNEL_PACKAGE)/boot
-	fiasco-gen -o $(CURDIR)/debian/$(KERNEL_PACKAGE)/boot/zImage-$(RELEASE).fiasco -g -k $(KSRC)/arch/arm/boot/zImage -v $(RELEASE)
+	test -f sdk-fiasco-gen_2.8.2.0.1+0m5_i386.deb || wget http://repository.maemo.org/pool/fremantle/non-free/s/sdk-fiasco-gen/sdk-fiasco-gen_2.8.2.0.1+0m5_i386.deb
+	ar x sdk-fiasco-gen_2.8.2.0.1+0m5_i386.deb data.tar.gz
+	tar -xf data.tar.gz ./usr/bin/fiasco-gen -O > fiasco-gen && chmod +x fiasco-gen
+	./fiasco-gen -o $(CURDIR)/debian/$(KERNEL_PACKAGE)/boot/zImage-$(RELEASE).fiasco -g -k $(KSRC)/arch/arm/boot/zImage -v $(RELEASE)
 	chmod 644 $(CURDIR)/debian/$(KERNEL_PACKAGE)/boot/zImage-$(RELEASE).fiasco
 
 install-bootimg:
@@ -107,9 +112,9 @@ install-debug: install-modules
 	mkdir -p $(CURDIR)/debian/$(DEBUG_PACKAGE)/lib/modules/$(RELEASE)
 	install -m644 $(KSRC)/vmlinux $(CURDIR)/debian/$(DEBUG_PACKAGE)/usr/lib/debug/vmlinux-$(RELEASE)
 	for f in $$(find $(CURDIR)/debian/$(MODULES_PACKAGE) -name "*.ko"); do \
-		objcopy --only-keep-debug $$f $(CURDIR)/debian/$(DEBUG_PACKAGE)/usr/lib/debug/lib/modules/$(RELEASE)/$$(basename $$f); \
-		strip --remove-section=.comment --remove-section=.note --strip-unneeded  $(CURDIR)/debian/$(MODULES_PACKAGE)/lib/modules/$(RELEASE)/$$(basename $$f); \
-		objcopy --add-gnu-debuglink $(CURDIR)/debian/$(DEBUG_PACKAGE)/usr/lib/debug/lib/modules/$(RELEASE)/$$(basename $$f) $(CURDIR)/debian/$(MODULES_PACKAGE)/lib/modules/$(RELEASE)/$$(basename $$f); \
+		arm-none-linux-gnueabi-objcopy --only-keep-debug $$f $(CURDIR)/debian/$(DEBUG_PACKAGE)/usr/lib/debug/lib/modules/$(RELEASE)/$$(basename $$f); \
+		arm-none-linux-gnueabi-strip --remove-section=.comment --remove-section=.note --strip-unneeded  $(CURDIR)/debian/$(MODULES_PACKAGE)/lib/modules/$(RELEASE)/$$(basename $$f); \
+		arm-none-linux-gnueabi-objcopy --add-gnu-debuglink $(CURDIR)/debian/$(DEBUG_PACKAGE)/usr/lib/debug/lib/modules/$(RELEASE)/$$(basename $$f) $(CURDIR)/debian/$(MODULES_PACKAGE)/lib/modules/$(RELEASE)/$$(basename $$f); \
 	done
 
 install-modules-debug: install-debug
@@ -126,7 +131,7 @@ install-modules:
 	mkdir -p $(CURDIR)/debian/$(MODULES_PACKAGE)/etc/modprobe.d
 	for f in $$(find $(KSRC) -name "*.ko"); do \
 		install -c -m644 $$f $(CURDIR)/debian/$(MODULES_PACKAGE)/lib/modules/$(RELEASE)/$$(basename $$f); \
-		strip --remove-section=.comment --remove-section=.note --strip-unneeded  $(CURDIR)/debian/$(MODULES_PACKAGE)/lib/modules/$(RELEASE)/$$(basename $$f); \
+		arm-none-linux-gnueabi-strip --remove-section=.comment --remove-section=.note --strip-unneeded  $(CURDIR)/debian/$(MODULES_PACKAGE)/lib/modules/$(RELEASE)/$$(basename $$f); \
 	done
 	install -m644 $(CURDIR)/debian/blacklist $(CURDIR)/debian/$(MODULES_PACKAGE)/etc/modprobe.d
 
@@ -219,9 +224,6 @@ endif
 	dh_testroot
 	dh_installchangelogs -s
 	dh_installdocs -s
-ifeq ($(ARCH),arm)
-	dh_strip -p$(HEADERS_PACKAGE)
-endif
 	dh_compress -s
 	dh_fixperms -s --exclude /usr/src/kernel-headers
 	dh_installdeb -s
