diff --git a/kernel-power-2.6.28/debian/control b/kernel-power-2.6.28/debian/control
index ac43331..ec1431c 100644
--- a/kernel-power-2.6.28/debian/control
+++ b/kernel-power-2.6.28/debian/control
@@ -3,7 +3,7 @@ Section: utils
 Priority: optional
 Maintainer: Pali Rohár <pali.rohar@gmail.com>
 XSBC-Bugtracker: https://garage.maemo.org/tracker/?group_id=1528
-Build-Depends: debhelper (>= 4.0.0), quilt, sdk-fiasco-gen, libncurses5-dev, sharutils
+Build-Depends: debhelper (>= 4.0.0), quilt, libncurses5-dev, sharutils
 Build-Depends-Indep: bzip2
 Standards-Version: 3.8.0
 
diff --git a/kernel-power-2.6.28/debian/patches/series b/kernel-power-2.6.28/debian/patches/series
index 6f53deb..8917eb7 100644
--- a/kernel-power-2.6.28/debian/patches/series
+++ b/kernel-power-2.6.28/debian/patches/series
@@ -4,6 +4,7 @@ nokia-20100903+0m5.diff
 nokia-20101501+0m5.diff
 nokia-20103103+0m5.diff
 2.6.28.10.diff
+madde.diff
 rx51_defconfig.diff
 unionfs-2.5.3.diff
 dm-loop.diff
diff --git a/kernel-power-2.6.28/debian/rules b/kernel-power-2.6.28/debian/rules
index 6510f9e..528c89b 100755
--- a/kernel-power-2.6.28/debian/rules
+++ b/kernel-power-2.6.28/debian/rules
@@ -86,7 +86,10 @@ install-kernel:
 	dh_testdir
 	dh_testroot
 	mkdir -p $(CURDIR)/debian/$(KERNEL_PACKAGE)/boot
-	fiasco-gen -o $(CURDIR)/debian/$(KERNEL_PACKAGE)/boot/zImage-$(RELEASE).fiasco -g -k $(KSRC)/arch/arm/boot/zImage -v $(RELEASE)
+	test -f sdk-fiasco-gen_2.8.2.0.1+0m5_i386.deb || wget http://repository.maemo.org/pool/fremantle/non-free/s/sdk-fiasco-gen/sdk-fiasco-gen_2.8.2.0.1+0m5_i386.deb
+	ar x sdk-fiasco-gen_2.8.2.0.1+0m5_i386.deb data.tar.gz
+	tar -xf data.tar.gz ./usr/bin/fiasco-gen -O > fiasco-gen && chmod +x fiasco-gen
+	./fiasco-gen -o $(CURDIR)/debian/$(KERNEL_PACKAGE)/boot/zImage-$(RELEASE).fiasco -g -k $(KSRC)/arch/arm/boot/zImage -v $(RELEASE)
 	chmod 644 $(CURDIR)/debian/$(KERNEL_PACKAGE)/boot/zImage-$(RELEASE).fiasco
 
 install-bootimg:
@@ -227,9 +230,6 @@ endif
 	dh_testroot
 	dh_installchangelogs -s
 	dh_installdocs -s
-ifeq ($(ARCH),arm)
-	dh_strip -p$(HEADERS_PACKAGE)
-endif
 	dh_compress -s
 	dh_fixperms -s --exclude /usr/src/kernel-headers
 	dh_installdeb -s
