diff --git a/kernel-power-2.6.28/debian/control b/kernel-power-2.6.28/debian/control
index b4f0406..cd80594 100644
--- a/kernel-power-2.6.28/debian/control
+++ b/kernel-power-2.6.28/debian/control
@@ -3,7 +3,7 @@ Section: utils
 Priority: optional
 Maintainer: Pali Rohár <pali.rohar@gmail.com>
 XSBC-Bugtracker: https://garage.maemo.org/tracker/?group_id=1528
-Build-Depends: debhelper (>= 4.0.0), quilt, sdk-fiasco-gen, libncurses5-dev, sharutils, wl1251-maemo-source
+Build-Depends: debhelper (>= 4.0.0), quilt, libncurses5-dev, sharutils
 Build-Depends-Indep: bzip2
 Standards-Version: 3.8.0
 
diff --git a/kernel-power-2.6.28/debian/patches/series b/kernel-power-2.6.28/debian/patches/series
index 469626f..43e30ca 100644
--- a/kernel-power-2.6.28/debian/patches/series
+++ b/kernel-power-2.6.28/debian/patches/series
@@ -9,6 +9,7 @@ nokia-20101501+0m5.diff
 nokia-20103103+0m5.diff
 #nokia-20103103+0m5_usb.diff
 2.6.28.10.diff
+madde.diff
 rx51_defconfig.diff
 unionfs-2.5.3.diff
 dm-loop.diff
diff --git a/kernel-power-2.6.28/debian/rules b/kernel-power-2.6.28/debian/rules
index a78394b..25397d5 100755
--- a/kernel-power-2.6.28/debian/rules
+++ b/kernel-power-2.6.28/debian/rules
@@ -1,6 +1,11 @@
 #!/usr/bin/make -f
 #
 
+MADDE_VERSION := $(shell ls $(HOME)/.madde/ | sort -n -r | head -n1)
+MADDE_TOOLCHAIN := $(shell ls $(HOME)/.madde/$(MADDE_VERSION)/toolchains/ | sort -n -r | head -n1)
+MADDE_PATH := $(HOME)/.madde/$(MADDE_VERSION)/toolchains/$(MADDE_TOOLCHAIN)/bin
+export PATH := $(MADDE_PATH):$(PATH)
+
 RELEASE := $(shell dpkg-parsechangelog | sed -ne 's/^Version: 1:\(.*\)/\1/p' | sed 's/-10/.10-/')
 REVISION := $(shell echo "$(RELEASE)" | sed 's/.*\./\./')
 EXTRAVERSION := EXTRAVERSION=$(REVISION)
@@ -134,9 +139,9 @@ install-debug: install-modules
 	mkdir -p $(CURDIR)/debian/$(DEBUG_PACKAGE)/lib/modules/$(RELEASE)
 	install -m644 $(KSRC)/vmlinux $(CURDIR)/debian/$(DEBUG_PACKAGE)/usr/lib/debug/vmlinux-$(RELEASE)
 	for f in $$(find $(CURDIR)/debian/$(MODULES_PACKAGE)/lib/modules/$(RELEASE) -name "*.ko"); do \
-		objcopy --only-keep-debug $$f $(CURDIR)/debian/$(DEBUG_PACKAGE)/usr/lib/debug/lib/modules/$(RELEASE)/$$(basename $$f); \
-		strip --remove-section=.comment --remove-section=.note --strip-unneeded  $(CURDIR)/debian/$(MODULES_PACKAGE)/lib/modules/$(RELEASE)/$$(basename $$f); \
-		objcopy --add-gnu-debuglink $(CURDIR)/debian/$(DEBUG_PACKAGE)/usr/lib/debug/lib/modules/$(RELEASE)/$$(basename $$f) $(CURDIR)/debian/$(MODULES_PACKAGE)/lib/modules/$(RELEASE)/$$(basename $$f); \
+		arm-none-linux-gnueabi-objcopy --only-keep-debug $$f $(CURDIR)/debian/$(DEBUG_PACKAGE)/usr/lib/debug/lib/modules/$(RELEASE)/$$(basename $$f); \
+		arm-none-linux-gnueabi-strip --remove-section=.comment --remove-section=.note --strip-unneeded  $(CURDIR)/debian/$(MODULES_PACKAGE)/lib/modules/$(RELEASE)/$$(basename $$f); \
+		arm-none-linux-gnueabi-objcopy --add-gnu-debuglink $(CURDIR)/debian/$(DEBUG_PACKAGE)/usr/lib/debug/lib/modules/$(RELEASE)/$$(basename $$f) $(CURDIR)/debian/$(MODULES_PACKAGE)/lib/modules/$(RELEASE)/$$(basename $$f); \
 	done
 
 install-modules-debug: install-debug
@@ -154,12 +159,12 @@ install-modules:
 	mkdir -p $(CURDIR)/debian/$(MODULES_PACKAGE)/opt/packet-injection-modules/$(RELEASE)
 	for f in $$(find $(KSRC) -name "*.ko"); do \
 		install -c -m644 $$f $(CURDIR)/debian/$(MODULES_PACKAGE)/lib/modules/$(RELEASE)/$$(basename $$f); \
-		strip --remove-section=.comment --remove-section=.note --strip-unneeded  $(CURDIR)/debian/$(MODULES_PACKAGE)/lib/modules/$(RELEASE)/$$(basename $$f); \
+		arm-none-linux-gnueabi-strip --remove-section=.comment --remove-section=.note --strip-unneeded  $(CURDIR)/debian/$(MODULES_PACKAGE)/lib/modules/$(RELEASE)/$$(basename $$f); \
 	done
 	for ko in wl1251_spi.ko wl1251.ko cfg80211.ko compat.ko rfkill_backport.ko mac80211.ko; do \
 		for f in $$(find $(COMPAT_WIRELESS_TREE) -name $$ko); do \
 			install -c -m644 $$f $(CURDIR)/debian/$(MODULES_PACKAGE)/opt/packet-injection-modules/$(RELEASE)/$$(basename $$f); \
-			strip --remove-section=.comment --remove-section=.note --strip-unneeded  $(CURDIR)/debian/$(MODULES_PACKAGE)/opt/packet-injection-modules/$(RELEASE)/$$(basename $$f); \
+			arm-none-linux-gnueabi-strip --remove-section=.comment --remove-section=.note --strip-unneeded  $(CURDIR)/debian/$(MODULES_PACKAGE)/opt/packet-injection-modules/$(RELEASE)/$$(basename $$f); \
 		done; \
 	done
 	install -m644 $(CURDIR)/debian/kernel-power.blacklist $(CURDIR)/debian/$(MODULES_PACKAGE)/etc/modprobe.d
@@ -253,9 +258,6 @@ endif
 	dh_testroot
 	dh_installchangelogs -s
 	dh_installdocs -s
-ifeq ($(ARCH),arm)
-	dh_strip -p$(HEADERS_PACKAGE)
-endif
 	dh_compress -s
 	dh_fixperms -s --exclude /usr/src/kernel-headers
 	dh_installdeb -s
