diff --git kernel-power-2.6.28/debian/control kernel-power-2.6.28/debian/control
index 9b9e930..578e377 100644
--- kernel-power-2.6.28/debian/control
+++ kernel-power-2.6.28/debian/control
@@ -3,7 +3,7 @@ Section: utils
 Priority: optional
 Maintainer: Pali Rohár <pali.rohar@gmail.com>
 XSBC-Bugtracker: https://garage.maemo.org/tracker/?group_id=1528
-Build-Depends: debhelper (>= 4.0.0), quilt, sdk-fiasco-gen, libncurses5-dev, sharutils
+Build-Depends: debhelper (>= 4.0.0), quilt, libncurses5-dev, sharutils
 Standards-Version: 3.8.0
 
 Package: kernel-power-flasher
diff --git kernel-power-2.6.28/debian/patches/series kernel-power-2.6.28/debian/patches/series
index 4ae2339..93a7c2d 100644
--- kernel-power-2.6.28/debian/patches/series
+++ kernel-power-2.6.28/debian/patches/series
@@ -35,3 +35,4 @@ l2cap_parent.patch
 wl12xx-rx-fix.diff
 vanilla-2.6.28-anti-io-stalling.patch
 joikuspot.diff
+madde.patch
diff --git kernel-power-2.6.28/debian/rules kernel-power-2.6.28/debian/rules
index ec876fa..e6f0a1c 100755
--- kernel-power-2.6.28/debian/rules
+++ kernel-power-2.6.28/debian/rules
@@ -101,7 +101,10 @@ install-kernel:
 	dh_installdirs
 
 	install -d -m 755 $(CURDIR)/debian/$(KERNEL_PACKAGE)/boot
-	fiasco-gen -o $(CURDIR)/debian/$(KERNEL_PACKAGE)/boot/zImage-$(RELEASE).fiasco -g \
+	test -f sdk-fiasco-gen_2.8.2.0.1+0m5_i386.deb || wget http://repository.maemo.org/pool/fremantle/non-free/s/sdk-fiasco-gen/sdk-fiasco-gen_2.8.2.0.1+0m5_i386.deb
+	ar x sdk-fiasco-gen_2.8.2.0.1+0m5_i386.deb data.tar.gz
+	tar -xf data.tar.gz ./usr/bin/fiasco-gen -O > fiasco-gen; chmod +x fiasco-gen
+	./fiasco-gen -o $(CURDIR)/debian/$(KERNEL_PACKAGE)/boot/zImage-$(RELEASE).fiasco -g \
 			-k $(BUILDTMP)/arch/arm/boot/zImage -v $(RELEASE)
 	chmod 644 $(CURDIR)/debian/$(KERNEL_PACKAGE)/boot/zImage-$(RELEASE).fiasco
 
@@ -134,12 +137,12 @@ install-headers:
 	dh_installdirs
 
 	install -d -m755 $(CURDIR)/debian/$(HEADERS_PACKAGE)/usr/src/$(HEADERS_PACKAGE)/
-	install -p -o root -g root -m 644 $(BUILDTMP)/.config  $(CURDIR)/debian/$(HEADERS_PACKAGE)/usr/src/$(HEADERS_PACKAGE)/
-	install -p -o root -g root -m 644 $(KSRC)/Makefile $(CURDIR)/debian/$(HEADERS_PACKAGE)/usr/src/$(HEADERS_PACKAGE)/
-	install -p -o root -g root -m 644 $(BUILDTMP)/Module.symvers $(CURDIR)/debian/$(HEADERS_PACKAGE)/usr/src/$(HEADERS_PACKAGE)/
-	cd $(BUILDTMP) && install -p -o root -g root -m 644 -D include/config/auto.conf \
+	install -p -m 644 $(BUILDTMP)/.config  $(CURDIR)/debian/$(HEADERS_PACKAGE)/usr/src/$(HEADERS_PACKAGE)/
+	install -p -m 644 $(KSRC)/Makefile $(CURDIR)/debian/$(HEADERS_PACKAGE)/usr/src/$(HEADERS_PACKAGE)/
+	install -p -m 644 $(BUILDTMP)/Module.symvers $(CURDIR)/debian/$(HEADERS_PACKAGE)/usr/src/$(HEADERS_PACKAGE)/
+	cd $(BUILDTMP) && install -p -m 644 -D include/config/auto.conf \
 		$(CURDIR)/debian/$(HEADERS_PACKAGE)/usr/src/$(HEADERS_PACKAGE)/include/config/auto.conf
-	cd $(BUILDTMP) && install -p -o root -g root -m 644 -D include/config/kernel.release \
+	cd $(BUILDTMP) && install -p -m 644 -D include/config/kernel.release \
 		$(CURDIR)/debian/$(HEADERS_PACKAGE)/usr/src/$(HEADERS_PACKAGE)/include/config/kernel.release
 	cd $(KSRC) && \
 		find arch/$(ARCH) $(foreach dir,$(INCLUDE_HEADERS),include/$(dir)) -name '*.h' \
@@ -238,9 +241,6 @@ endif
 	dh_installchangelogs -s
 	dh_installdocs -s
 	dh_install -s
-ifeq ($(ARCH),arm)
-	dh_strip -p$(HEADERS_PACKAGE)
-endif
 	dh_compress -s
 	dh_fixperms -s --exclude /usr/src/kernel-headers
 	dh_installdeb -s
