--- kernel-power-2.6.28/Makefile	2011-07-09 10:24:38.656013905 +0200
+++ kernel-power-2.6.28/Makefile	2011-07-09 10:24:40.536013905 +0200
@@ -4,6 +4,8 @@ SUBLEVEL = 28
 EXTRAVERSION = .10
 NAME = Erotic Pickled Herring
 
+export PATH := /usr/bin:$(HOME)/.madde/0.7.48/toolchains/arm-2007q3-51sb6-gdb71-arm-none-linux-gnueabi_linux_x86_64/bin:$(PATH)
+
 # *DOCUMENTATION*
 # To see a list of typical targets execute "make help"
 # More info can be located in ./README
@@ -196,7 +198,7 @@ SUBARCH := arm
 # Note: Some architectures assign CROSS_COMPILE in their arch/*/Makefile
 export KBUILD_BUILDHOST := $(SUBARCH)
 ARCH		?= $(SUBARCH)
-CROSS_COMPILE	?= arm-linux-
+CROSS_COMPILE	?= arm-none-linux-gnueabi-
 
 # Architecture as present in compile.h
 UTS_MACHINE 	:= $(ARCH)
--- kernel-power-2.6.28/scripts/unifdef.c	2011-01-23 18:22:03.000000000 +0100
+++ kernel-power-2.6.28/scripts/unifdef.c	2011-01-23 18:22:17.000000000 +0100
@@ -206,6 +206,7 @@
 static void             error(const char *);
 static int              findsym(const char *);
 static void             flushline(bool);
+#define getline _getline
 static Linetype         getline(void);
 static Linetype         ifeval(const char **);
 static void             ignoreoff(void);
