--- kernel-cssu/drivers/power/power_supply_sysfs.c	2012-01-28 02:03:53.273467047 +0100
+++ kernel-cssu/drivers/power/power_supply_sysfs.c	2012-01-28 02:04:20.085466262 +0100
@@ -60,8 +60,8 @@ static ssize_t power_supply_show_propert
 
 	if (ret < 0) {
 		if (ret != -ENODEV)
-			dev_err(dev, "driver failed to report `%s' property\n",
-				attr->attr.name);
+/*			dev_err(dev, "driver failed to report `%s' property\n",
+				attr->attr.name);*/
 		return ret;
 	}
 
