Index: kernel-cssu-2.6.28/arch/arm/mach-omap2/board-rx51.c
===================================================================
--- kernel-cssu-2.6.28.orig/arch/arm/mach-omap2/board-rx51.c	2012-05-25 04:12:52.708683352 -0400
+++ kernel-cssu-2.6.28/arch/arm/mach-omap2/board-rx51.c	2012-05-25 03:52:23.617140000 -0400
@@ -45,6 +45,8 @@
 #include <mach/prcm.h>
 #include "cm.h"
 
+#include "board-rx51-secure.h"
+
 #define RX51_USB_TRANSCEIVER_RST_GPIO	67
 
 extern int omap_init_fb(void);
@@ -162,6 +164,56 @@
 	.set_pm_limits	= rx51_usb_set_pm_limits,
 };
 
+#ifdef CONFIG_ARM_ERRATA_430973
+static ssize_t rx51_errata_430973_show(struct kobject *kobj,
+				       struct kobj_attribute *attr,
+				       char *buf)
+{
+	u32 acr;
+      
+	/* Read ACR */
+	asm volatile ("mrc p15, 0, %0, c1, c0, 1" : "=r" (acr));
+
+	return sprintf(buf,"%u\n",(acr >> 6) & 1);
+}
+
+static ssize_t rx51_errata_430973_store(struct kobject *kobj,
+					struct kobj_attribute *attr,
+					const char *buf, size_t n)
+{
+	u32 value;
+
+	if (sscanf(buf, "%u", &value) != 1) 
+	{
+		printk(KERN_ERR "errata_430973: Invalid value\n");
+		return -EINVAL;
+	}
+	if(value)
+		rx51_secure_update_aux_cr(1 << 6, 0);
+	else
+		rx51_secure_update_aux_cr(0, 1 << 6);
+
+	return n;
+}
+
+static struct kobj_attribute rx51_errata_430973 = {
+	.attr = {
+	.name = __stringify(errata_430973),
+	.mode = 0644,
+	},
+	.show = rx51_errata_430973_show,
+	.store = rx51_errata_430973_store,
+};
+
+static void rx51_add_errata_430973_sysfs(void)
+{
+	int ret = 0;
+	ret = sysfs_create_file(kernel_kobj, &rx51_errata_430973.attr);
+	if (ret)
+		printk(KERN_ERR "sysfs_create_file failed: %d\n", ret);
+}
+#endif
+
 static void __init rx51_init(void)
 {
 	rx51_xceiv_init();
@@ -176,6 +228,13 @@
 	 */
 	omap_init_fb();
 	omap_bt_init(&rx51_bt_config);
+
+#ifdef CONFIG_ARM_ERRATA_430973
+	printk(KERN_INFO "Enabling ARM errata 430973 workaround.\n");
+	/* set IBE to 1 */
+	rx51_secure_update_aux_cr(1 << 6, 0);
+	rx51_add_errata_430973_sysfs();
+#endif
 }
 
 static void __init rx51_map_io(void)
