--- kernel-power/drivers/power/Makefile	2012-04-29 18:03:10.751810073 +0200
+++ kernel-power/drivers/power/Makefile	2012-04-29 18:03:52.207808860 +0200
@@ -25,3 +25,4 @@ obj-$(CONFIG_TWL4030_BCI_BATTERY)	+= twl
 obj-$(CONFIG_BATTERY_TOSA)	+= tosa_battery.o
 obj-$(CONFIG_BATTERY_WM97XX)	+= wm97xx_battery.o
 obj-$(CONFIG_BATTERY_BQ27x00)	+= bq27x00_battery.o
+obj-$(CONFIG_CHARGER_BQ2415X)	+= bq2415x_charger.o
--- kernel-power/drivers/power/Kconfig	2012-04-29 18:01:48.819812466 +0200
+++ kernel-power/drivers/power/Kconfig	2012-04-29 18:03:06.043810209 +0200
@@ -75,4 +75,10 @@ config BATTERY_BQ27x00
 	help
 	  Say Y here to enable support for batteries with BQ27200(I2C) chip.
 
+config CHARGER_BQ2415X
+	tristate "BQ2415X charger driver"
+	depends on I2C
+	help
+	  Say Y here to enable support for chargers with BQ2415X(I2C) chip.
+
 endif # POWER_SUPPLY
