--- kernel-power-2.6.28.orig/arch/arm/mach-omap2/pm.c	2011-11-23 11:17:07.555970502 +0000
+++ kernel-power-2.6.28/arch/arm/mach-omap2/pm.c	2011-11-23 11:10:32.626825000 +0000
@@ -322,6 +322,8 @@
 			resource_lock_opp(PRCM_VDD1);
 			vdd1_locked = 1;
 		}
+		else
+			return n;
 	} else if (attr == &vdd2_lock_attr) {
 		flags = OPP_IGNORE_LOCK;
 		attr = &vdd2_opp_attr;
@@ -335,6 +337,8 @@
 			resource_lock_opp(PRCM_VDD2);
 			vdd2_locked = 1;
 		}
+		else
+			return n;
 	}
 
 	if (attr == &vdd1_opp_attr) {
