--- kernel-maemo-2.6.28.orig/drivers/mtd/devices/block2mtd.c
+++ kernel-maemo-2.6.28/drivers/mtd/devices/block2mtd.c
@@ -285,7 +285,8 @@
 
 	dev->mtd.size = dev->blkdev->bd_inode->i_size & PAGE_MASK;
 	dev->mtd.erasesize = erase_size;
-	dev->mtd.writesize = 1;
+	dev->mtd.writesize = 2048;
+	dev->mtd.subpage_sft = 2;
 	dev->mtd.type = MTD_RAM;
 	dev->mtd.flags = MTD_CAP_RAM;
 	dev->mtd.erase = block2mtd_erase;
