--- kernel-2.6.28.orig/Makefile
+++ kernel-2.6.28/Makefile
@@ -1,7 +1,7 @@
 VERSION = 2
 PATCHLEVEL = 6
 SUBLEVEL = 28
-EXTRAVERSION =
+EXTRAVERSION = maemo
 NAME = Erotic Pickled Herring
 
 # *DOCUMENTATION*
--- kernel-2.6.28.orig/scripts/kconfig/lxdialog/check-lxdialog.sh
+++ kernel-2.6.28/scripts/kconfig/lxdialog/check-lxdialog.sh
@@ -68,12 +68,14 @@
 		check
 		;;
 	"-ccflags")
-		ccflags
+		#ccflags
+		echo '-DCURSES_LOC="<ncurses.h>"'
 		;;
 	"-ldflags")
 		shift
 		cc="$@"
-		ldflags
+		#ldflags
+		echo "-lncurses"
 		;;
 	"*")
 		usage
