--- kernel-maemo-2.6.28.orig/net/mac80211/rc80211_minstrel.c
+++ kernel-maemo-2.6.28/net/mac80211/rc80211_minstrel.c
@@ -224,7 +224,7 @@
 	unsigned int sample_ndx;
 	sample_ndx = SAMPLE_TBL(mi, mi->sample_idx, mi->sample_column);
 	mi->sample_idx++;
-	if (mi->sample_idx > (mi->n_rates - 2)) {
+	if (mi->n_rates == 1 || mi->sample_idx > (mi->n_rates - 2)) {
 		mi->sample_idx = 0;
 		mi->sample_column++;
 		if (mi->sample_column >= SAMPLE_COLUMNS)
