--- kernel-power-2.6.28/arch/arm/mach-omap2/board-rx51-peripherals.c	2011-01-23 03:53:13.000000000 +0100
+++ kernel-power-2.6.28/arch/arm/mach-omap2/board-rx51-peripherals.c	2011-01-23 04:02:21.000000000 +0100
@@ -572,6 +580,9 @@ static struct i2c_board_info __initdata 
 		I2C_BOARD_INFO("tsl2563", 0x29),
 		.platform_data = &rx51_tsl2563_platform_data,
 	},
+	{
+		I2C_BOARD_INFO("bq27200", 0x55),
+	},
 };
 
 static struct i2c_board_info __initdata rx51_peripherals_i2c_board_info_3[] = {
