--- kernel-maemo-2.6.28.orig/scripts/kconfig/lxdialog/check-lxdialog.sh
+++ kernel-maemo-2.6.28/scripts/kconfig/lxdialog/check-lxdialog.sh
@@ -68,12 +68,14 @@
 		check
 		;;
 	"-ccflags")
-		ccflags
+		#ccflags
+		echo '-DCURSES_LOC="<ncurses.h>"'
 		;;
 	"-ldflags")
 		shift
 		cc="$@"
-		ldflags
+		#ldflags
+		echo "-lncurses"
 		;;
 	"*")
 		usage
--- kernel-maemo-2.6.28.orig/drivers/gpu/pvr/Makefile
+++ kernel-maemo-2.6.28/drivers/gpu/pvr/Makefile
@@ -21,5 +21,5 @@
 			   -DPVR_BUILD_DIR="\"$(PVR_BUILD_DIR)\""	\
 			   -DPVR_BUILD_DATE="\"$(DATE)\""
 
-ccflags-y		+= $(CBUILD) -include $(src)/pvrconfig.h
+ccflags-y		+= $(CBUILD) -include $(KBUILD_SRC)/$(src)/pvrconfig.h
 
