--- kernel-power-2.6.28.orig/drivers/mmc/card/block.c
+++ kernel-power-2.6.28/drivers/mmc/card/block.c
@@ -479,7 +479,7 @@
 	struct mmc_blk_data *md;
 	int devidx, ret;
 
-	devidx = find_first_zero_bit(dev_use, MMC_NUM_MINORS);
+	devidx = find_next_zero_bit(dev_use, MMC_NUM_MINORS, card->host->mmcblk_devidx);
 	if (devidx >= MMC_NUM_MINORS)
 		return ERR_PTR(-ENOSPC);
 	__set_bit(devidx, dev_use);
--- kernel-power-2.6.28.orig/drivers/mmc/host/omap_hsmmc.c
+++ kernel-power-2.6.28/drivers/mmc/host/omap_hsmmc.c
@@ -1759,8 +1759,8 @@
 	omap_hsmmc_protect_card(host);
 
 	mmc_add_host(mmc);
-
 	if (mmc_slot(host).name != NULL) {
+		if (mmc_slot(host).name[0]=='e') mmc->mmcblk_devidx=1; /* 'e'xternal slot should start at mmcblk1 */
 		ret = device_create_file(&mmc->class_dev, &dev_attr_slot_name);
 		if (ret < 0)
 			goto err_slot_name;
--- kernel-power-2.6.28.orig/include/linux/mmc/host.h
+++ kernel-power-2.6.28/include/linux/mmc/host.h
@@ -207,7 +207,7 @@
 #endif
 
 	struct dentry		*debugfs_root;
-
+	unsigned int		mmcblk_devidx; /* preferred mmc block device index (mmcblkX) */
 	unsigned long		private[0] ____cacheline_aligned;
 };
 
