diff -ruNd kernel-2.6.28.old/arch/arm/mach-omap2/smartreflex.c kernel-2.6.28/arch/arm/mach-omap2/smartreflex.c
--- kernel-2.6.28.old/arch/arm/mach-omap2/smartreflex.c	2010-05-25 09:13:49.000000000 +0200
+++ kernel-2.6.28/arch/arm/mach-omap2/smartreflex.c	2010-05-25 09:08:37.000000000 +0200
@@ -1004,6 +1004,11 @@
 		return -EINVAL;
 	}
 
+	if (value != 0) {
+		pr_warning("VDD2 smartreflex is broken\n");
+		return -EINVAL;
+	}
+
 	mutex_lock(&dvfs_mutex);
 
 	current_vdd2opp_no = resource_get_level("vdd2_opp");
