--- kernel-power-2.6.28.orig/drivers/usb/gadget/nokia.c
+++ kernel-power-2.6.28/drivers/usb/gadget/nokia.c
@@ -142,8 +142,7 @@
 	.bind		= nokia_bind_config,
 	.bConfigurationValue = 1,
 	/* .iConfiguration = DYNAMIC */
-	.bmAttributes	= USB_CONFIG_ATT_ONE,
-	.bMaxPower	= 250, /* 500mA */
+	.bmAttributes		= USB_CONFIG_ATT_SELFPOWER
 };
 
 static struct usb_configuration nokia_config_100ma_driver = {
@@ -151,8 +150,7 @@
 	.bind		= nokia_bind_config,
 	.bConfigurationValue = 2,
 	/* .iConfiguration = DYNAMIC */
-	.bmAttributes	= USB_CONFIG_ATT_ONE | USB_CONFIG_ATT_SELFPOWER,
-	.bMaxPower	= 50, /* 100 mA */
+	.bmAttributes		= USB_CONFIG_ATT_SELFPOWER
 };
 
 static int __init nokia_bind(struct usb_composite_dev *cdev)
--- kernel-power-2.6.28.orig/arch/arm/mach-omap2/Makefile
+++ kernel-power-2.6.28/arch/arm/mach-omap2/Makefile
@@ -100,7 +100,8 @@
 					   board-rx51-peripherals.o \
 					   mmc-twl4030.o \
 					   ssi.o \
-					   usb-musb.o
+					   usb-ehci.o \
+					   usb-musb.o 
 obj-$(CONFIG_MACH_NOKIA_RX71)		+= board-rx71.o \
 					   board-rx71-peripherals.o
 
