--- kernel-power-2.6.28.orig/arch/arm/mach-omap2/prcm.c
+++ kernel-power-2.6.28/arch/arm/mach-omap2/prcm.c
@@ -131,7 +131,7 @@
  * HACK for RX51 boards previous to B3 which
  * doesn't have a reset line to isp1707 transceiver
  */
-extern void musb_emergency_stop(void);
+/*extern void musb_emergency_stop(void);*/
 
 /* Resets clock rates and reboots the system. Only called from system.h */
 void omap_prcm_arch_reset(char mode)
@@ -156,7 +156,7 @@
 	 * HACK for RX51 boards previous to B3 which
 	 * doesn't have a reset line to isp1707 transceiver
 	 */
-	musb_emergency_stop();
+	/*musb_emergency_stop();*/
 
 	prm_set_mod_reg_bits(OMAP_RST_GS, prcm_offs, RM_RSTCTRL);
 }
--- kernel-power-2.6.28.orig/drivers/i2c/chips/twl4030-poweroff.c
+++ kernel-power-2.6.28/drivers/i2c/chips/twl4030-poweroff.c
@@ -39,6 +39,7 @@
 {
 	twl4030_usb_suspended = suspended;
 }
+EXPORT_SYMBOL(twl4030_upd_usb_suspended);
 
 static void twl4030_poweroff(void)
 {
